pkg/target/http.go

```diff
@@ -215,63 +337,150 @@ func AdaptHTTPTargetFunc(f func(c *HTTPTargetConfig) (*HTTPTarget, error)) HTTPT
+			for _, msg := range goodMsgs {
+				msg.TimeRequestStarted = requestStarted
+				msg.TimeRequestFinished = requestFinished
+			}
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error sending request: "+err.Error()))
+				continue
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+				for _, msg := range goodMsgs {
+					if msg.AckFunc != nil { // Ack successful messages
+						msg.AckFunc()
+					}
+					sent = append(sent, msg)
+				}
+				continue
+			}
+
+			responseBody, err := io.ReadAll(resp.Body)
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error reading response body: "+err.Error()))
+				continue
+			}
+
+			response := response{Body: string(responseBody), Status: resp.StatusCode}
+
+			if findMatchingRule(response, ht.responseRules.Invalid) != nil {
+				for _, msg := range goodMsgs {
+					// can we use response body as an error message for invalid data?
+					msg.SetError(errors.New(response.Body))
+				}
+
+				invalid = append(invalid, goodMsgs...)
+				continue
+			}
+
+			var errorDetails error
+			if rule := findMatchingRule(response, ht.responseRules.SetupError); rule != nil {
+				hitSetupError = true
+				if rule.MatchingBodyPart != "" {
+					errorDetails = fmt.Errorf("Got setup error, response status: '%s' with error details: '%s'", resp.Status, rule.MatchingBodyPart)
+				} else {
+					errorDetails = fmt.Errorf("Got setup error, response status: '%s'", resp.Status)
+				}
+			} else {
+				errorDetails = fmt.Errorf("Got transient error, response status: '%s'", resp.Status)
+			}
+			errResult = multierror.Append(errResult, errorDetails)
+			failed = append(failed, goodMsgs...)
 		}
 	}
-	if errResult != nil {
-		errResult = errors.Wrap(errResult, "Error sending http requests")
+
+	if hitSetupError {
+		errResult = models.SetupWriteError{Err: errResult}
 	}
 
 	ht.log.Debugf("Successfully wrote %d/%d messages", len(sent), len(messages))
-	return models.NewTargetWriteResult(
-		sent,
-		failed,
-		oversized,
-		invalid,
-	), errResult
+	return models.NewTargetWriteResult(sent, failed, oversized, invalid), errResult
+}
+
+func findMatchingRule(res response, rules []Rule) *Rule {
+	for _, rule := range rules {
+		if ruleMatches(res, rule) {
+			return &rule
+		}
+	}
+	return nil
+}
+
+func ruleMatches(res response, rule Rule) bool {
+	codeMatch := httpStatusMatches(res.Status, rule.MatchingHTTPCodes)
+	if rule.MatchingBodyPart != "" {
+		return codeMatch && responseBodyMatches(res.Body, rule.MatchingBodyPart)
+	}
+	return codeMatch
+}
+
+func httpStatusMatches(actual int, expectedCodes []int) bool {
+	for _, expected := range expectedCodes {
+		if expected == actual {
+			return true
+		}
+	}
+	return false
+}
+
+func responseBodyMatches(actual string, bodyPattern string) bool {
+	return strings.Contains(actual, bodyPattern)
```

Comment:
I'm not sure if this came up in review of this feature. Just calling out explicitly that I think this implementation works for the present requirements, but we might encounter requirements to support regex or pattern matching in future.